lxd/cgroup/init.go

```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
 	return "unknown"
 }
 
+// Resources is a generic type used to abstract resource control features
+// support for the legacy and unified hierarchy.
+type Resources int
+
+const (
+	// Blkio resource control
+	Blkio Resources = iota
+	// BlkioWeight resource control
+	BlkioWeight
+	// CPU resource control
+	CPU
+	// CPUAcct resource control
+	CPUAcct
+	// CPUSet resource control
+	CPUSet
+	// Devices resource control
+	Devices
+	// Freezer resource control
+	Freezer
+	// Memory resource control
+	Memory
+	// MemoryMaxUsage resource control
+	MemoryMaxUsage
+	// MemorySwap resource control
+	MemorySwap
+	// MemorySwapMaxUsage resource control
+	MemorySwapMaxUsage
+	// MemorySwapUsage resource control
+	MemorySwapUsage
+	// MemorySwappiness resource control
+	MemorySwappiness
+	// NetPrio resource control
+	NetPrio
+	// Pids resource control
+	Pids
+)
+
+// SupportsVersion indicates whether or not a given cgroup control knob is
+// available and in which type of cgroup filesystem.
+func (info *Info) SupportsVersion(resource Resources) (Backend, bool) {
+	switch resource {
+	case Blkio:
+		val, ok := cgControllers["blkio"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case BlkioWeight:
+		val, ok := cgControllers["blkio.weight"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPU:
+		val, ok := cgControllers["cpu"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPUAcct:
+		val, ok := cgControllers["cpuacct"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case CPUSet:
+		val, ok := cgControllers["cpuset"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case Devices:
+		val, ok := cgControllers["devices"]
+		return val, ok
```

Comment:
In such cases, I think `return cgControllers["devices"]` is actually cleaner

Reply:
Assuming Go does the right thing in such cases, maps are a bit magic (can be one or two values dependent on context).

Reply:
That doesn't work for maps...

Reply:
```go
# github.com/lxc/lxd/lxd/cgroup
lxd/cgroup/init.go:159:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:161:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:163:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:165:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
```

Reply:
Yeah, was afraid of that, Go could do with a tiny bit more smarts around that case.